File: c-buffer/src/main/c/endianness.c
/**
 * Copyright (c) 2022 by Kristoffer Paulsson <kristoffer.paulsson@talenten.se>.
 *
 * This software is available under the terms of the MIT license. Parts are licensed
 * under different terms if stated. The legal terms are attached to the LICENSE file
 * and are made available on:
 *
 *      https://opensource.org/licenses/MIT
 *
 * SPDX-License-Identifier: MIT
 *
 * Contributors:
 *      Kristoffer Paulsson - initial implementation
 */
#include "endianness.h"

enum {
    kLittleEndian = 2,
    kBigEndian = 1,
    kUnknownEndian = 0
};

int endian() {
#if ENDIAN_ORDER == _LITTLE_ENDIAN
    return kLittleEndian;
#elif ENDIAN_ORDER == _BIG_ENDIAN
    return kBigEndian;
#elif ENDIAN_ORDER==_PDP_ENDIAN
    return kUnknownEndian;
#else
    return kUnknownEndian;
#endif
}

void speedmemcpy(void *dest, const void * src, uint32_t n) {
#if INTPTR_MAX == INT64_MAX
    uint32_t big = n / sizeof(uint64_t);
    uint32_t small = n % sizeof(uint64_t);

    uint64_t *dest_big = (uint64_t *) dest;
    uint64_t *src_big = (uint64_t *) src;

    for (uint32_t i = 0; i < big; i++) {
        dest_big[i] = src_big[i];
    }

    char *dest_small = (char *) dest + big * sizeof(uint64_t);
    char *src_small = (char *) src + big * sizeof(uint64_t);
#elif INTPTR_MAX == INT32_MAX
    uint32_t big = n / sizeof(uint32_t);
    uint32_t small = n % sizeof(uint32_t);

    uint32_t *dest_big = (uint32_t *) dest;
    uint32_t *src_big = (uint32_t *) src;

    for (uint32_t i = 0; i < big; i++) {
        dest_big[i] = src_big[i];
    }

    char *dest_small = (char *) dest + big * sizeof(uint32_t);
    char *src_small = (char *) src + big * sizeof(uint32_t);
#else
#error "Environment must be 64-bit or 32-bit"
#endif
    for (uint32_t j = 0; j < small; j++) {
        dest_small[j] = src_small[j];
    }
}

File: buffer/src/commonMain/kotlin/org/angelos/io/buf/Endianness.kt
/**
 * Copyright (c) 2022 by Kristoffer Paulsson <kristoffer.paulsson@talenten.se>.
 *
 * This software is available under the terms of the MIT license. Parts are licensed
 * under different terms if stated. The legal terms are attached to the LICENSE file
 * and are made available on:
 *
 *      https://opensource.org/licenses/MIT
 *
 * SPDX-License-Identifier: MIT
 *
 * Contributors:
 *      Kristoffer Paulsson - initial implementation
 */
package org.angelos.io.buf

/**
 * Representing endianness in systems and data buffers or streams.
 *
 * @property endian
 * @constructor Create empty Endianness
 */
enum class Endianness(val endian: Boolean) {

    /**
     * Big endian enumeration value.
     *
     * @constructor Create empty Big Endian
     */
    BIG_ENDIAN(false),

    /**
     * Little endian enumeration value.
     *
     * @constructor Create empty Little Endian
     */
    LITTLE_ENDIAN(true);

    /**
     * Checks if the endianness is big.
     *
     * @return
     */
    fun isBig(): Boolean = endian == BIG_ENDIAN.endian

    /**
     * Checks if the endianness is little.
     *
     * @return
     */
    fun isLittle(): Boolean = endian == LITTLE_ENDIAN.endian

    companion object {

        /**
         * Native order of the running environment.
         *
         * @return the endian
         */
        fun nativeOrder(): Endianness = when (Internals.getEndian()) {
            1 -> BIG_ENDIAN
            2 -> LITTLE_ENDIAN
            else -> throw BufferException("Unknown type of endian.")
        }
    }

    /**
     * String representation of the endianness.
     *
     * @return human readable endianness
     */
    override fun toString(): String = when (endian) {
        false -> "BIG_ENDIAN"
        true -> "LITTLE_ENDIAN"
    }
}